### PR TITLE
Add slot methods to a module instead of the component class itself

### PR DESCRIPTION
**Pull request by ozydingo**
### What are you trying to accomplish?

Addresses #2027:

- Define slot methods in a dynamic module so that slot name methods can be overridden with ability to call `super`

### What approach did you choose and why?

- Inspired by ActiveRecord's naming of `GeneratedAttributeMethods`, I am choosing the name `GeneratedSlotMethods`
- Define a new module and set it to the const `self::GeneratedSlotMethods` in the `inherited` hook defined by `ViewComponent::Slotable`, similar to how ActiveRecord creates `<model>::GeneratedAttributeMethods`.
- include `self::GeneratedSlotMethods` in the same inherited hook
- define the `<slot>` and `<slot>?` methods on the module

### TODO

- [x] FIX: Use the `inherited` hook and not the `included` hook; `ViewComponent::Slotable` is included only once by `ViewComponent::Base`
- [x] Add test to make sure independent components do not share methods
- [x] Ensure component subclasses that define their own slots also inherit superclass slot methods
- [x] Investigate other slottable dynamic methods (currently commented out)

### Anything you want to highlight for special attention from reviewers?

- ~Added a step in CONTRIBUTING to `bundle install`, with the goal of installing appraisal (otherwise `bundle exec appraisal install` would fail)~
- Running `bundle exec appraisal install` modified the gemfiles and I'm not sure why. I have no experience with appraisal. I imagine I will need to revert these changes, ~but I'm keeping them on the branch for now since if I revert I can't run tests, and also to ask about it.~

### CODE REVIEW

**Comment by ozydingo** [comment]
@joelhawksley after a "brief" hiatus I cam back to this and got it in working order. Unfortunately the "Test compatibility with Primer ViewComponents" is failing, and if anyone had some bandwidth to help shine some light into that failure that would be hugely appreciated. I'll try to dig in when I get another chance but I'll be wandering in a lot of unfamiliar territory as I do! It's also pretty unclear to me, at a high level, why this change would specifically break the `test_single_select_item_checked_via_keyboard_space` test so I suspect the resolution will involve a lot of digging into those component specifics.

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -42,14 +42,15 @@ The codespace environment includes a minimal Rails app with ViewComponent instal
...
-8. Wait for the pull request to be reviewed and merged.
+1. Configure and install local dependencies: `bundle install`.
+1. Configure and install testing dependencies: `bundle exec appraisal install`.
+1. Make sure the tests pass: `bundle exec appraisal rake` (see below for specific cases).
+1. Create a new branch: `git checkout -b my-branch-name`.
+1. Add tests, make the change, and make sure the tests still pass.
+1. Add an entry to the top of `docs/CHANGELOG.md` for the changes, no matter how small.
+1. If it's your first time contributing, add yourself to `docs/index.md`.
+1. Push to the fork and [submit a pull request](https://github.com/viewcomponent/view_component/compare).
+1. Wait for the pull request to be reviewed and merged.
```
Comment (joelhawksley) [inline docs/CONTRIBUTING.md:53]:
```suggestion
1. Configure and install the dependencies: `bundle exec appraisal install`.
2. Make sure the tests pass: `bundle exec appraisal rake` (see below for specific cases).
3. Create a new branch: `git checkout -b my-branch-name`.
4. Add tests, make the change, and make sure the tests still pass.
5. Add an entry to the top of `docs/CHANGELOG.md` for the changes, no matter how small.
6. If it's your first time contributing, add yourself to `docs/index.md`.
7. Push to the fork and [submit a pull request](https://github.com/viewcomponent/view_component/compare).
8. Wait for the pull request to be reviewed and merged.
```

**Review by joelhawksley — COMMENTED** [review]
Looks good to me! I'm going to defer to @BlakeWilliams for the review here ❤️

**Review by BlakeWilliams — APPROVED** [review]
LGTM! Thanks for this improvement!

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,10 @@ nav_order: 5
 
 ## main
 
+* Allow overridden slot methods to use `super` to call the standard implementation.
```
Comment (joelhawksley) [inline docs/CHANGELOG.md:13]:
```suggestion
* Allow overridden slot methods to use `super`.
```

**Pull request merged**